doc/04-Upgrading.md

```diff
@@ -3,6 +3,44 @@
 Specific version upgrades are described below. Please note that version upgrades are incremental.
 If you are upgrading across multiple versions, make sure to follow the steps for each of them.
 
+## Upgrading to Icinga DB v1.1.2
+
+Please apply the `1.1.2.sql` upgrade script to your database.
+For package installations, you can find this file at `/usr/share/icingadb/schema/mysql/upgrades/` or
+`/usr/share/icingadb/schema/pgsql/upgrades/`, depending on your database type.
+
+As the daemon checks the schema version, the recommended way to perform the upgrade is to stop the daemon, apply the
+schema upgrade and then start the new daemon version. If you want to minimize downtime as much as possible, it is safe
+to apply this schema upgrade while the Icinga DB v1.1.1 daemon is still running and then restart the daemon with the
+new version. Please keep in mind that depending on the distribution, your package manager may automatically attempt to
+restart the daemon when upgrading the package.
+
+### Upgrading the state_history Table (optional, [#655](https://github.com/Icinga/icingadb/issues/655))
+
+Applying `1.1.2.sql` itself will allow you to set hosts/services' `max_check_attempts` above 255.
+But actual soft state changes with check attempt above 255 in your history will still crash Icinga DB v1.1.2.
+To avoid that, you can lower `max_check_attempts`. Optionally increase the respective `retry_interval`
+not to change the timespan from an outage to a hard state (`max_check_attempts x retry_interval`).
+
+If lower `max_check_attempts` are not an option, you'll have to execute the SQL statements below.
+Note that this can take a lot of time depending on the history size and the performance of the database.
+While the upgrade is running, that table will be locked and can't be accessed.
+This means that the existing history can't be viewed in Icinga Web and new history entries will be buffered in Redis.
+Do not interrupt them not to corrupt the state_history table!
```

Comment:
This should give a bit of context for the reader. Simply starting with a sentence that earlier versions weren't able to handle `max_check_attempts > 255` should be enough.

> If lower `max_check_attempts` are not an option, you'll have to execute the SQL statements below.

This makes it sound like not applying the optional part of the upgrade was the preferred way. I don't think it is. Instead, it's a trade-off where both options have pros and cons. Thus, it should be presented as such so that users are encouraged to make an informed decision here.

I think it's also worth mentioning that the optional part can be applied at a later time, for example during the next longer maintenance window.

> Do not interrupt them not to corrupt the state_history table!

This won't corrupt the table (unless there's a bug in the database server), it might just result in an expensive rollback that could make everything take much longer than necessary.

Reply:
I don't agree that this is clearly the less dangerous operation. The trade-off is between performing a somewhat expensive maintenance operation right now, or do I want to put the burden on myself to never ever configure `max_check_attempts > 255` because if I do, it will break. You might not be able to know the exact duration the schema upgrade will take but you know when it will happen, so it's annoying but predictable. Skipping the upgrade and being an innocent-looking config option away from crashing Icinga DB in the future, which I'd call dangerous.

Reply:
Then I opt for presenting the history table upgrade as preferred with the duration disclaimer.

Reply:
Yes, I think the following key points would appropriately describe the what optional means here:

- History table used too small data type, fixing it is expensive, so we don't do it automatically on upgrade.
- If applying the fix is fine for you, you should do it.
- If not, you have to ensure no `max_check_attempts > 255` now or in the future (if you can't, reconsider the previous step).
- Applying at a later time also works, but until then there must be no `max_check_attempts > 255`.